core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go

```diff
@@ -0,0 +1,349 @@
+			r.lggr.Error("[OffchainLookup] buildCallOpts:", err)
+			continue
+		}
+		// need upkeep info for offchainConfig and to hit callback
+		upkeepInfo, err := r.getUpkeepInfo(upkeepId, err, opts)
+		if err != nil {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_OFFCHAIN_LOOKUP_ERROR
+			r.lggr.Error("[OffchainLookup] GetUpkeep:", err)
+			continue
+		}
+
+		// 	do the http request
+		body, statusCode, err := r.doRequest(offchainLookup, upkeepId, upkeepInfo.OffchainConfig)
+		if err != nil {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_OFFCHAIN_LOOKUP_ERROR
+			r.lggr.Error("[OffchainLookup] doRequest:", err)
+			continue
+		}
+		r.lggr.Debugf("[OffchainLookup] StatusCode: %d\n", statusCode)
+		//r.lggr.Debugf("[OffchainLookup] Body: %s\n", string(body))
+
+		values, err := offchainLookup.parseJson(body)
+		if err != nil {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_OFFCHAIN_LOOKUP_ERROR
+			r.lggr.Error("[OffchainLookup] parseJson:", err)
+			continue
+		}
+		r.lggr.Debugf("[OffchainLookup] Parsed values: %+v\n", values)
+
+		needed, performData, err := r.offchainLookupCallback(ctx, offchainLookup, values, statusCode, upkeepInfo, opts)
+		if err != nil {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_OFFCHAIN_LOOKUP_ERROR
+			r.lggr.Error("[OffchainLookup] offchainLookupCallback=", err)
+			continue
+		}
+		if !needed {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_UPKEEP_NOT_NEEDED
+			r.lggr.Debug("[OffchainLookup] callback reports upkeep not needed")
+			continue
+		}
+
+		// success!
+		upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_NONE
+		upkeepResults[i].State = types.Eligible
+		upkeepResults[i].PerformData = performData
+		r.lggr.Debugf("[OffchainLookup] Success: %+v\n", upkeepResults[i])
+	}
+	return upkeepResults, nil
+}
+
+func (r *EvmRegistry) getUpkeepInfo(upkeepId *big.Int, err error, opts *bind.CallOpts) (keeper_registry_wrapper2_0.UpkeepInfo, error) {
+	zero := common.Address{}
+	var upkeepInfo keeper_registry_wrapper2_0.UpkeepInfo
+	u, found := r.upkeepCache.Get(upkeepId.String())
+	if found {
+		upkeepInfo = u.(keeper_registry_wrapper2_0.UpkeepInfo)
+		r.lggr.Debugf("[OffchainLookup] cache hit UpkeepInfo: %+v\n", upkeepInfo)
+	} else {
+		upkeepInfo, err = r.registry.GetUpkeep(opts, upkeepId)
+		if err != nil {
+			return upkeepInfo, err
+		}
+		if upkeepInfo.Target == zero {
+			return upkeepInfo, errors.New("upkeepInfo should not be nil")
+		}
+		r.lggr.Debugf("[OffchainLookup] cache miss UpkeepInfo: %+v\n", upkeepInfo)
+		r.upkeepCache.Set(upkeepId.String(), upkeepInfo, cache.DefaultExpiration)
+	}
+	return upkeepInfo, nil
+}
+
+// decodeOffchainLookup decodes the revert error ChainlinkAPIFetch(string url, bytes extraData, string[] jsonFields, bytes4 callbackSelector)
+func (r *EvmRegistry) decodeOffchainLookup(data []byte) (OffchainLookup, error) {
+	offchainLookup := OffchainLookup{}
```

Comment:
No big deal but normally I would drop this line, and then get rid of lines 155-158, and instead just instantiate the type's fields when returning, e.g.

```
return OffchainLookup{
		url: *abi.ConvertType(errorParameters[0], new(string)).(*string),
		extraData: *abi.ConvertType(errorParameters[1], new([]byte)).(*[]byte),
		fields: *abi.ConvertType(errorParameters[2], new([]string)).(*[]string),
		callbackFunction: *abi.ConvertType(errorParameters[3], new([4]byte)).(*[4]byte)
	}, nil
```